circup/commands.py

```diff
@@ -406,44 +406,57 @@ def install(
 
 @main.command()
 @click.option("--overwrite", is_flag=True, help="Overwrite the file if it exists.")
+@click.option("--list", is_flag=True, help="List available examples.")
 @click.argument(
-    "examples", required=True, nargs=-1, shell_complete=completion_for_example
+    "examples", required=False, nargs=1, shell_complete=completion_for_example
 )
 @click.pass_context
-def example(ctx, examples, overwrite):
+# pylint: disable=redefined-builtin
+def example(ctx, list, examples, overwrite):
     """
-    Copy named example(s) from a bundle onto the device. Multiple examples
-    can be installed at once by providing more than one example name, each
-    separated by a space.
+    Copy named example from a bundle onto the device.
+    Option --list will show available examples.  --list will also
+    support partial searches.
     """
 
-    for example_arg in examples:
-        available_examples = get_bundle_examples(
-            get_bundles_list(), avoid_download=True
-        )
-        if example_arg in available_examples:
-            filename = available_examples[example_arg].split(os.path.sep)[-1]
+    if list and not examples:
+        click.echo("\n".join(completion_for_example(ctx, "", "")))
+        return
 
-            if overwrite or not ctx.obj["backend"].file_exists(filename):
-                click.echo(
-                    f"{'Copying' if not overwrite else 'Overwriting'}: {filename}"
-                )
-                ctx.obj["backend"].install_module_py(
-                    {"path": available_examples[example_arg]}, location=""
-                )
-            else:
-                click.secho(
-                    f"File: {filename} already exists. Use --overwrite if you wish to replace it.",
-                    fg="red",
-                )
+    if list:
+        click.echo("\n".join(completion_for_example(ctx, "", examples)))
+        return
+
+    available_examples = get_bundle_examples(get_bundles_list(), avoid_download=True)
+
+    if examples in available_examples:
+        install_metadata = {"path": available_examples[examples]}
+
+        # check of we are dealing with a file that needs to be pushed as code.py
+        # or a directory that needs to be copied raw
+        filename = available_examples[examples].split(os.path.sep)[-1]
+        if os.path.isfile(available_examples[examples]):
+            filename = "code.py"
```

Comment:
I think it would good to retain the ability to push the example with it's original name. I don't necessarily have a strong preference about which way is default, but if anyone has gotten used to the workflow involving the examples with their existing name it would be disruptive to have it now only allow the code.py naming. 

If it defaults to `code.py` I think we should add some `--no-rename` or similar flag that will make it keep the original name. Or vice versa we could keep the current behavior as default and add a `--code` something to let the user specify to rename it to code.py.